File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.py
import sys
sys.setrecursionlimit(10**6)
def solution(user_id, banned_id):
    #만족하는 문자열
    #banned_id에 따른 일치하는 문자열 리스트로 정리
    visited = set()
    check = [0 for _ in range(len(user_id))]
    def recur(x):
        nonlocal check
        if x >= len(banned_id):
            tmp = tuple(check)
            if tmp not in visited:
                visited.add(tmp)
                return
            return
        
        for u in range(len(user_id)):
            cur_user = user_id[u]
            #문자열 일치
            if len(cur_user) == len(banned_id[x]):
                cnt = 0
                cur = banned_id[x]
                for i in range(len(cur_user)):
                    if cur_user[i] == cur[i] or cur[i] == '*':
                        # print(cnt)
                        cnt += 1
                    else:
                        break
                # print(cnt,len(cur_user))
                if cnt == len(cur_user) and check[u]==0:
                    check[u] = 1
                    recur(x+1)
                    check[u] = 0
        return
    recur(0)
    print(len(visited))
    answer = 0
    return len(visited)

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.java
import java.util.*;
class Solution {
    static Set<String> set;
    static Map<String, List<String>> map;
    static Set<String> visited;
    static Set<Set<String>> visited_pair;
    static int ans;
    public int solution(String[] user_id, String[] banned_id) {
        set = new HashSet<>();
        map = new HashMap<>();
        visited = new HashSet<>();
        visited_pair = new HashSet<>();
        ans =0;
        for(int i=0;i<banned_id.length;i++){
            set.add(banned_id[i]);
        }
        for(String s:set){
            for(int i=0;i<user_id.length;i++){
                //매치되면 map에 추가
                if(match(user_id[i],s)){
                    List<String> tmp = map.getOrDefault(s,new ArrayList<>());
                    tmp.add(user_id[i]);
                    map.put(s,tmp);
                }
            }
        }
        dfs(0,banned_id);
        return ans;
    }
    public boolean match(String user_id,String pattern){
        if(user_id.length() != pattern.length()){
            return false;
        }
        else{
            for(int i=0;i<user_id.length();i++){
                if(pattern.charAt(i) != '*' 
                   &&user_id.charAt(i) != pattern.charAt(i)){
                    return false;
                }
            }
            return true;
        }
    }
    public void dfs(int idx, String[] banned_id){
        if(visited.size() == banned_id.length ){
            Set<String> tmp = new HashSet<>(visited);
            if(!visited_pair.contains(tmp)){
                visited_pair.add(tmp);
                ans += 1;
            }
            return;
        }
        else{
            for(String s:map.get(banned_id[idx])){
                if(!visited.contains(s)){
                    visited.add(s);
                    dfs(idx+1, banned_id);
                    visited.remove(s);
                }
            }
        }
    }
}